torch/csrc/distributed/c10d/Backend.cpp

```diff
@@ -18,6 +29,9 @@ void commonEventinit(
   evt.sequence_number = work.getSequencenumber();
   evt.operation = c10d::opTypeToString(work.retrieveOpType());
   evt.drop_count = 0;
+  // isCompleted is mutable :facepalm:
+  if (const_cast<Work&>(work).isCompleted() && !work.isSuccess())
```

Comment:
@fduwjj I think we should discuss this part

we're calling work.exception() which will query some cuda APIs.

we're doing this from the main watchdog thread.

i think that in today's watchdog design this is already something we're doing in the main loop, so i don't have a big reservation about it.

but if you're redesigning the loops to put cuda accesses on another thread, you'd also want to move this part.  would be good to think through how we'll handle the callback reporting error message safely, or else maybe not add the error to the callback in the first place.  (I think it should be possible to do though)